### ISSUE TITLE
can't use const as attribute value

### ISSUE THREAD

**Issue by srijs** [rust-lang/rust#52393]
I'm using serde to implement deserialization for a custom protocol. To keep the code dry, I've pulled all relevant strings into constants, but that creates a problem when trying to use the constants as attributes values in serde's derive macro.

Here's a playground with a minimal example to reproduce the problem: http://play.rust-lang.org/?gist=47230ae9671b24b312bef9540e520ddf&version=stable&mode=debug&edition=2015

The check that seems to produce the error is part of `libsyntax`: https://github.com/rust-lang/rust/blob/0db03e635a5e38ebc7635637b870b8fbcc8a7e46/src/libsyntax/parse/attr.rs#L261-L263

With my very basic knowledge of the rust compiler, I can see two potential ways this could be made possible:

1) Run derive macros after the constants have been expanded. This would be transparent to the macros (they could treat the expanded constants as literals), however this might be a no-go due to the way that the passes are ordered within rustc?

2) Relax the check in `libsyntax` (and possibly other places) to allow for identifiers to be valid attribute values on the syntax level, so that custom derive macros could use the identifier as part of their code gen (this wouldn't work for all macros, but at least for serde it looks like it'd be good enough).

If possible, could you give an indication as to whether this is something that you'd like in the compiler (in which case I'd be interested in working on a PR or RFC for it), or whether you think this is either impossible or undesirable?

**Comment by webern** [comment]
I'm surprised this issue doesn't have more upvotes. Is this issue represented somewhere else where it has more of a following. What I want to do is this:

```rust
const SOME_CONST: &str = "foobar";

#[some_derive_macro(name = SOME_CONST)]
struct Whatever {

}
```

**Comment by thomcc** [comment]
Yeah, you can't put consts in attributes, but it would be very nice. I'm not sure if solving this requires an RFC, but it might.

**Comment by oldgalileo** [comment]
The `include_str!` macro seems related, but I'm curious if there's been any work unlinked here that enables the usage of a `const _: &str` in an attribute yet?

**Comment by glyh** [comment]
Potential use case: [pomsky](https://pomsky-lang.org/) + [logos](https://github.com/maciejhirsz/logos)

**Comment by glyh** [comment]
PR-welcomed?
